lib/psbt.ex

```diff
@@ -195,6 +280,138 @@ defmodule Bitcoinex.PSBT.In do
+
+    val_len = TxUtils.serialize_compact_size_unsigned_int(byte_size(val))
+    key_len <> key <> val_len <> val
+  end
+
+  defp serialize_kv(:sighash_type, value) when value != nil do
+    key = <<@psbt_in_partial_sig::big-size(8)>>
+    key_len = TxUtils.serialize_compact_size_unsigned_int(byte_size(key))
+    val = <<value::little-size(32)>>
+    val_len = TxUtils.serialize_compact_size_unsigned_int(byte_size(val))
+    key_len <> key <> val_len <> val
+  end
+
+  defp serialize_kv(:final_scriptsig, value) when value != nil do
+    key = <<@psbt_in_final_scriptsig::big-size(8)>>
+    key_len = TxUtils.serialize_compact_size_unsigned_int(byte_size(key))
+    val = Base.decode16!(value, case: :lower)
+    val_len = TxUtils.serialize_compact_size_unsigned_int(byte_size(val))
+    key_len <> key <> val_len <> val
+  end
+
+  defp serialize_kv(:redeem_script, value) when value != nil do
+    key = <<@psbt_in_redeem_script::big-size(8)>>
+    key_len = TxUtils.serialize_compact_size_unsigned_int(byte_size(key))
+    val = Base.decode16!(value, case: :lower)
+    val_len = TxUtils.serialize_compact_size_unsigned_int(byte_size(val))
+    key_len <> key <> val_len <> val
+  end
+
+  defp serialize_kv(:witness_script, value) when value != nil do
+    key = <<@psbt_in_witness_script::big-size(8)>>
+    key_len = TxUtils.serialize_compact_size_unsigned_int(byte_size(key))
+    val = Base.decode16!(value, case: :lower)
+    val_len = TxUtils.serialize_compact_size_unsigned_int(byte_size(val))
+    key_len <> key <> val_len <> val
+  end
+
+  defp serialize_kv(:final_scriptwitness, value) when value != nil do
+    key = <<@psbt_in_final_scriptwitness::big-size(8)>>
+    key_len = TxUtils.serialize_compact_size_unsigned_int(byte_size(key))
+    val = Witness.serialize_witness([value])
+    val_len = TxUtils.serialize_compact_size_unsigned_int(byte_size(val))
+    key_len <> key <> val_len <> val
+  end
+
+  defp serialize_kv(:bip32_derivation, value) when value != nil do
+    key = <<@psbt_in_bip32_derivation::big-size(8)>>
+    key_data = Base.decode16!(value.public_key, case: :lower)
+
+    key = key <> key_data
+    key_len = TxUtils.serialize_compact_size_unsigned_int(byte_size(key))
+
+    val =
+      <<value.pfp::little-size(32)>> <>
+        (for(chunk <- value.derivation, do: <<chunk::little-size(32)>>)
+         |> :erlang.list_to_binary())
+
+    val_len = TxUtils.serialize_compact_size_unsigned_int(byte_size(val))
+    key_len <> key <> val_len <> val
+  end
+
+  defp serialize_kv(_key, _value) do
+    <<>>
+  end
+
+  def serialize_inputs(inputs) when is_list(inputs) and length(inputs) > 0 do
+    serialize_input(inputs, <<>>)
+  end
+
+  def serialize_inputs(_inputs) do
+    <<>>
+  end
+
+  defp serialize_input([], serialized_inputs), do: serialized_inputs
+
+  defp serialize_input(inputs, serialized_inputs) do
+    [input | inputs] = inputs
+
+    serialized_input =
+      serialize_kv(:non_witness_utxo, input.non_witness_utxo) <>
```

Comment:
Should we filter out nil value here so we don't need a catch all for the `serialize_kv`?
```
[:non_witness_utxo, .., :witness_script]
|> Enum.reduce(fn k, acc ->
  case Map.get(input, serialize_kv) do
    nil -> acc
    v -> acc <> serialize_kv(k, v)
  end
end)
```

My main concern is the need of considering nil as possible value in nested private functions.